### PR TITLE
Refactor : Manage metadata using a distributed K/V engine

### PR DESCRIPTION
**Pull request by xiaguan**
In order to ensure the fault tolerance and consistency of metadata, we plan to use distributed K/V engine to reconstruct metadata.

### CODE REVIEW

**Comment by ActivePeter** [comment]
Link the design issue

**Comment by xiaguan** [comment]
#369

File: src/async_fuse/memfs/kv_engine/mod.rs [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+                TxnCmp::Equal,
+                version,
+            );
+        }
+        // add the write operations
+        for (key, value) in &self.buffer {
+            let key = key.clone();
+            if let Some(ref value) = *value {
+                // TODO(xiaguan) : maybe too many clone?
+                let put_req = etcd_client::EtcdPutRequest::new(key.clone(), value.clone());
```
Comment (rogercloud) [inline src/async_fuse/memfs/kv_engine/mod.rs:167]: We can consider reference if it's not shared during this operation.

File: src/async_fuse/memfs/kv_engine/mod.rs [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+        let req = etcd_client::EtcdGetRequest::new(key.clone());
+        let mut resp = self
+            .client
+            .kv()
+            .get(req)
+            .await
+            .with_context(|| format!("failed to get GetResponse from etcd, key={key:?}"))?;
+        let kvs = resp.take_kvs();
+        // panic here because we don't expect to have multiple values for one key
+        assert!(kvs.len() <= 1, "multiple values for one key");
```
Comment (rogercloud) [inline src/async_fuse/memfs/kv_engine/mod.rs:124]: Return an error for bad response?

Comment (xiaguan) [reply src/async_fuse/memfs/kv_engine/mod.rs]: I added `assert` statements for the sake of convenience during development and debugging.They will not be effective in the release versionThese assert statements describe certain scenarios that are not supported by kv_engine, such as `path2inum(key)`. In these scenarios, the caller expects to receive either none or one value, but not two values. Even if the caller were to handle this situation as an error, they would not be able to fix it, as it is considered undefined behavior.

Comment (xiaguan) [reply src/async_fuse/memfs/kv_engine/mod.rs]: `panic` is comment typo error . I will fix it .

**Review by rogercloud — APPROVED** [review]

**Pull request merged**